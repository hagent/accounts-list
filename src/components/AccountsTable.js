import React, { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import TableSortFields from "./TableSortFields";
import {
  createSorter,
  toggleSortFieldInSortList,
  getOrderSign
} from "../models/sort";

const MainPanel = styled.div`
  display: flex;
  flex-direction: column;
`;

const TableHeader = styled.th`
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border-bottom: 3px solid black;
  text-align: left;
  padding: 5px 10px;
  cursor: pointer;
  &:hover {
    background: rgb(200, 200, 200);
  }
`;

const TableCell = styled.td`
  border-bottom: 1px solid black;
  padding: 5px 10px;
`;

const Table = styled.table`
  border-collapse: collapse;
  margin-top: 10px;
`;

const COLUMNS = [
  {
    field: "name",
    title: "Name"
  },
  {
    field: "profitLoss",
    title: "Profit & Loss"
  },
  {
    field: "accountTypeTitle",
    title: "Account Type"
  }
];

// abstract sortable table could be extracted from the component
// though I decided that without additional requirements it would be premature complication
export default function AccountsTable({ accounts }) {
  const [sortFields, setSortFields] = useState([]);

  const toggleSortField = fieldName => () =>
    setSortFields(toggleSortFieldInSortList(sortFields, fieldName));
  const removeSortField = fieldName => () =>
    setSortFields(sortFields.filter(x => x.field !== fieldName));
  const sort = createSorter(sortFields);
  const sortedAccounts = sort(accounts);
  return (
    <MainPanel>
      <TableSortFields
        sortFields={sortFields}
        toggleField={toggleSortField}
        removeField={removeSortField}
        columns={COLUMNS}
      />
      <Table>
        <thead>
          <tr>
            {COLUMNS.map(c => (
              <TableHeader key={c.field} onClick={toggleSortField(c.field)}>
                {c.title} {getOrderSign(sortFields, c.field)}
              </TableHeader>
            ))}
          </tr>
        </thead>
        <tbody>
          {sortedAccounts.map(a => (
            <tr key={a.id}>
              <TableCell>{a.name}</TableCell>
              <TableCell>
                {a.currency} {a.profitLoss}
              </TableCell>
              <TableCell>{a.accountTypeTitle}</TableCell>
            </tr>
          ))}
        </tbody>
      </Table>
    </MainPanel>
  );
}

AccountsTable.propTypes = {
  accounts: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      currency: PropTypes.string.isRequired,
      accountTypeTitle: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired
    })
  ).isRequired
};
