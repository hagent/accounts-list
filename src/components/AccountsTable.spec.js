import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import AccountsTable from "./AccountsTable";

describe("AccountsTable", () => {
  it("renders table", () => {
    const accounts = [
      {
        id: 1,
        name: "Spread bet",
        profitLoss: 0.23,
        accountType: "IGSB",
        currency: "$",
        accountTypeTitle: "Spread bet account"
      },
      {
        id: 2,
        name: "New Spread bet",
        profitLoss: -679,
        accountType: "IGSB",
        currency: "$",
        accountTypeTitle: "Spread bet account"
      }
    ];

    const wrapper = shallow(<AccountsTable accounts={accounts} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
