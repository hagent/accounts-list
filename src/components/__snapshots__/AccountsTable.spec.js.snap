// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsTable renders table 1`] = `
<styled.div>
  <TableSortFields
    columns={
      Array [
        Object {
          "field": "name",
          "title": "Name",
        },
        Object {
          "field": "profitLoss",
          "title": "Profit & Loss",
        },
        Object {
          "field": "accountTypeTitle",
          "title": "Account Type",
        },
      ]
    }
    removeField={[Function]}
    sortFields={Array []}
    toggleField={[Function]}
  />
  <styled.table>
    <thead>
      <tr>
        <styled.th
          key="name"
          onClick={[Function]}
        >
          Name
           
        </styled.th>
        <styled.th
          key="profitLoss"
          onClick={[Function]}
        >
          Profit & Loss
           
        </styled.th>
        <styled.th
          key="accountTypeTitle"
          onClick={[Function]}
        >
          Account Type
           
        </styled.th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="1"
      >
        <styled.td>
          Spread bet
        </styled.td>
        <styled.td>
          $
           
          0.23
        </styled.td>
        <styled.td>
          Spread bet account
        </styled.td>
      </tr>
      <tr
        key="2"
      >
        <styled.td>
          New Spread bet
        </styled.td>
        <styled.td>
          $
           
          -679
        </styled.td>
        <styled.td>
          Spread bet account
        </styled.td>
      </tr>
    </tbody>
  </styled.table>
</styled.div>
`;
